File: src/creatures/enemy.js
import * as PIXI from 'pixi.js'
import { Creature } from '../creature.js'
import { toCoords, toKey } from '../cave.js'

function manhattanDistance(a, b) {
    return Math.abs(a.x - b.x) + Math.abs(a.y - b.y)
}

function isEnemyCreature(creature) {
    return creature?.assignment === 'enemy' || creature?.constructor?.name === 'Enemy'
}

function isTrackedTrilobite(creature) {
    return creature?.constructor?.name === 'Trilobite'
}

export class Enemy extends Creature {

    constructor(name, location, game, sprite = new PIXI.Sprite("Enemy")) {
        super(name, location, sprite, game)
        this.assignment = 'enemy'
        this.enemyTargetTileKey = null
    }

    getBehavior() {
        return this.enemyBehavior
    }

    cleanupBeforeRemoval() {
        this.enemyTargetTileKey = null
    }

    enemyBehavior = () => {
        return this.enqueueAction(() => this.enemyStep1())
    }

    ensureEnemyState() {
        if (isEnemyCreature(this)) {
            return true
        }

        this.enemyTargetTileKey = null
        const fallbackBehavior = this.getBehavior()
        if (typeof fallbackBehavior === 'function' && fallbackBehavior !== this.enemyBehavior) {
            fallbackBehavior.call(this)
        }
        return false
    }

    clearEnemyTarget() {
        this.enemyTargetTileKey = null
    }

    getHostileTrilobites() {
        if (!this.cave) {
            return []
        }

        const hostiles = []
        for (const creature of this.cave.trilobites) {
            if (!creature || creature === this || !isTrackedTrilobite(creature)) {
                continue
            }

            hostiles.push(creature)
        }

        return hostiles
    }

    getHostileAtTileKey(tileKey) {
        if (!tileKey) {
            return null
        }

        for (const hostile of this.getHostileTrilobites()) {
            if (toKey(hostile.location) === tileKey) {
                return hostile
            }
        }

        return null
    }

    getHostileBuildingAtTileKey(tileKey) {
        if (!this.cave || !tileKey) {
            return null
        }

        const tile = this.cave.getTile(tileKey)
        if (!tile) {
            return null
        }

        const building = tile.getBuilt()
        if (!building || building.cave !== this.cave || building.health <= 0) {
            return null
        }

        return building
    }

    getHostileTargetAtTileKey(tileKey) {
        return this.getHostileAtTileKey(tileKey) ?? this.getHostileBuildingAtTileKey(tileKey)
    }

    isAdjacentToTileKey(tileKey, location = this.location) {
        const tileCoords = toCoords(tileKey)
        if (!Number.isFinite(tileCoords?.x) || !Number.isFinite(tileCoords?.y)) {
            return false
        }

        return manhattanDistance(location, tileCoords) === 1
    }

    getAdjacentHostileTileKey(location = this.location) {
        if (!this.cave) {
            return null
        }

        const currentTile = this.cave.getTile(toKey(location))
        if (!currentTile) {
            return null
        }

        let adjacentBuildingTileKey = null
        for (const neighbor of currentTile.getNeighbors()) {
            if (this.getHostileAtTileKey(neighbor.key)) {
                return neighbor.key
            }

            if (!adjacentBuildingTileKey && this.getHostileBuildingAtTileKey(neighbor.key)) {
                adjacentBuildingTileKey = neighbor.key
            }
        }

        return adjacentBuildingTileKey
    }

    enemyStep1 = () => {
        if (!this.ensureEnemyState()) {
            return false
        }

        if (this.enemyTargetTileKey && this.isAdjacentToTileKey(this.enemyTargetTileKey)) {
            return this.enemyStep2()
        }

        const adjacentHostileTileKey = this.getAdjacentHostileTileKey()
        if (adjacentHostileTileKey) {
            this.enemyTargetTileKey = adjacentHostileTileKey
            return this.enemyStep2()
        }

        return this.enemyStep3()
    }

    enemyStep2 = () => {
        if (!this.ensureEnemyState()) {
            return false
        }

        if (!this.enemyTargetTileKey) {
            return this.enemyStep3()
        }

        const hostile = this.getHostileTargetAtTileKey(this.enemyTargetTileKey)
        if (!hostile) {
            this.clearEnemyTarget()
            return this.enemyStep3()
        }

        if (!this.isAdjacentToTileKey(this.enemyTargetTileKey)) {
            return this.enemyStep3()
        }

        const dealt = this.dealDamage(hostile)
        if (!this.getHostileTargetAtTileKey(this.enemyTargetTileKey)) {
            this.clearEnemyTarget()
        }

        return dealt > 0
    }

    enemyStep3 = () => {
        if (!this.ensureEnemyState()) {
            return false
        }

        if (this.enemyTargetTileKey && !this.getHostileTargetAtTileKey(this.enemyTargetTileKey)) {
            this.clearEnemyTarget()
        }

        const nextLocation = this.cave?.getBfsFieldNextStep('colony', this.location)
        if (!nextLocation) {
            this.clearEnemyTarget()
            return false
        }

        this.clearActionQueue()
        this.pathPreview.push({ x: nextLocation.x, y: nextLocation.y })
        return this.enemyStepMove(nextLocation)
    }

    enemyStepMove = (nextLocation) => {
        if (!this.ensureEnemyState()) {
            return false
        }

        if (this.enemyTargetTileKey && !this.getHostileTargetAtTileKey(this.enemyTargetTileKey)) {
            this.clearEnemyTarget()
            this.clearActionQueue()
            return this.enemyStep3()
        }

        const adjacentHostileTileKey = this.getAdjacentHostileTileKey()
        if (adjacentHostileTileKey) {
            this.enemyTargetTileKey = adjacentHostileTileKey
            this.clearActionQueue()
            return this.enemyStep2()
        }

        const moved = this.performMove(nextLocation)
        if (moved === false) {
            this.clearActionQueue()
            return this.enemyStep3()
        }

        if (this.pathPreview.length > 0) {
            this.pathPreview.shift()
        }

        if (this.enemyTargetTileKey && this.isAdjacentToTileKey(this.enemyTargetTileKey)) {
            this.clearActionQueue()
            return this.enemyStep2()
        }

        const nextAdjacentHostileTileKey = this.getAdjacentHostileTileKey()
        if (nextAdjacentHostileTileKey) {
            this.enemyTargetTileKey = nextAdjacentHostileTileKey
            this.clearActionQueue()
            return this.enemyStep2()
        }

        return moved
    }

}

File: TriloGame.CSharp/src/TriloGame.Game/Core/Entities/Enemy.cs
using TriloGame.Game.Core.Simulation;
using TriloGame.Game.Shared.Math;

namespace TriloGame.Game.Core.Entities;

public sealed class Enemy : Creature
{
    public Enemy(string name, GridPoint location, GameSession session)
        : base(name, location, session)
    {
        Assignment = "enemy";
    }

    public string? EnemyTargetTileKey { get; private set; }

    public override Action? GetBehavior() => EnemyBehavior;

    public override void CleanupBeforeRemoval(object? source = null)
    {
        EnemyTargetTileKey = null;
    }

    private void EnemyBehavior()
    {
        EnqueueAction(() => EnemyStep1());
    }

    public bool EnsureEnemyState()
    {
        if (Assignment == "enemy")
        {
            return true;
        }

        EnemyTargetTileKey = null;
        var fallback = GetBehavior();
        if (fallback is not null && fallback.Method != ((Action)EnemyBehavior).Method)
        {
            fallback();
        }

        return false;
    }

    public void ClearEnemyTarget()
    {
        EnemyTargetTileKey = null;
    }

    public List<Trilobite> GetHostileTrilobites()
    {
        return Cave is null ? [] : Cave.Trilobites.ToList();
    }

    public Trilobite? GetHostileAtTileKey(string? tileKey)
    {
        return string.IsNullOrWhiteSpace(tileKey)
            ? null
            : GetHostileTrilobites().FirstOrDefault(hostile => hostile.Location.ToString() == tileKey);
    }

    public Core.Buildings.Building? GetHostileBuildingAtTileKey(string? tileKey)
    {
        if (Cave is null || string.IsNullOrWhiteSpace(tileKey))
        {
            return null;
        }

        var tile = Cave.GetTile(tileKey);
        var building = tile?.Built;
        return building is not null && building.Cave == Cave && building.Health > 0 ? building : null;
    }

    public object? GetHostileTargetAtTileKey(string? tileKey)
    {
        return (object?)GetHostileAtTileKey(tileKey) ?? GetHostileBuildingAtTileKey(tileKey);
    }

    public bool IsAdjacentToTileKey(string tileKey, GridPoint? location = null)
    {
        return GridPoint.ManhattanDistance(location ?? Location, GridPoint.Parse(tileKey)) == 1;
    }

    public string? GetAdjacentHostileTileKey(GridPoint? location = null)
    {
        var currentTile = Cave?.GetTile((location ?? Location).ToString());
        if (currentTile is null)
        {
            return null;
        }

        string? adjacentBuildingTileKey = null;
        foreach (var neighbor in currentTile.Neighbors)
        {
            if (GetHostileAtTileKey(neighbor.Key) is not null)
            {
                return neighbor.Key;
            }

            if (adjacentBuildingTileKey is null && GetHostileBuildingAtTileKey(neighbor.Key) is not null)
            {
                adjacentBuildingTileKey = neighbor.Key;
            }
        }

        return adjacentBuildingTileKey;
    }

    public bool EnemyStep1()
    {
        if (!EnsureEnemyState())
        {
            return false;
        }

        if (EnemyTargetTileKey is not null && IsAdjacentToTileKey(EnemyTargetTileKey))
        {
            return EnemyStep2();
        }

        var adjacent = GetAdjacentHostileTileKey();
        if (adjacent is not null)
        {
            EnemyTargetTileKey = adjacent;
            return EnemyStep2();
        }

        return EnemyStep3();
    }

    public bool EnemyStep2()
    {
        if (!EnsureEnemyState())
        {
            return false;
        }

        if (EnemyTargetTileKey is null)
        {
            return EnemyStep3();
        }

        var hostile = GetHostileTargetAtTileKey(EnemyTargetTileKey);
        if (hostile is null)
        {
            ClearEnemyTarget();
            return EnemyStep3();
        }

        if (!IsAdjacentToTileKey(EnemyTargetTileKey))
        {
            return EnemyStep3();
        }

        var dealt = DealDamage(hostile);
        if (GetHostileTargetAtTileKey(EnemyTargetTileKey) is null)
        {
            ClearEnemyTarget();
        }

        return dealt > 0;
    }

    public bool EnemyStep3()
    {
        if (!EnsureEnemyState())
        {
            return false;
        }

        if (EnemyTargetTileKey is not null && GetHostileTargetAtTileKey(EnemyTargetTileKey) is null)
        {
            ClearEnemyTarget();
        }

        var nextLocation = Cave?.GetBfsFieldNextStep("colony", Location);
        if (nextLocation is null)
        {
            ClearEnemyTarget();
            return false;
        }

        ClearActionQueue();
        PathPreview.Add(nextLocation.Value);
        return EnemyStepMove(nextLocation.Value);
    }

    public bool EnemyStepMove(GridPoint nextLocation)
    {
        if (!EnsureEnemyState())
        {
            return false;
        }

        if (EnemyTargetTileKey is not null && GetHostileTargetAtTileKey(EnemyTargetTileKey) is null)
        {
            ClearEnemyTarget();
            ClearActionQueue();
            return EnemyStep3();
        }

        var adjacent = GetAdjacentHostileTileKey();
        if (adjacent is not null)
        {
            EnemyTargetTileKey = adjacent;
            ClearActionQueue();
            return EnemyStep2();
        }

        var moved = PerformMove(nextLocation);
        if (!moved)
        {
            ClearActionQueue();
            return EnemyStep3();
        }

        if (PathPreview.Count > 0)
        {
            PathPreview.RemoveAt(0);
        }

        if (EnemyTargetTileKey is not null && IsAdjacentToTileKey(EnemyTargetTileKey))
        {
            ClearActionQueue();
            return EnemyStep2();
        }

        var nextAdjacent = GetAdjacentHostileTileKey();
        if (nextAdjacent is not null)
        {
            EnemyTargetTileKey = nextAdjacent;
            ClearActionQueue();
            return EnemyStep2();
        }

        return moved;
    }
}
